Mathlib/NumberTheory/Zagier.lean

```diff
@@ -0,0 +1,278 @@
+def groupPowers : Group (powers f) where
+  inv := fun ⟨g, hg⟩ => ⟨g^[p - 1], by
+    rw [mem_powers_iff] at hg ⊢
+    obtain ⟨k, hk⟩ := hg
+    use k * (p - 1)
+    rw [← pow_eq_iterate, ← hk, pow_mul]⟩
+  mul_left_inv := fun ⟨g, hg⟩ => by
+    simp only [ge_iff_le, mk_mul_mk]
+    congr
+    rw [← pow_eq_iterate, ← pow_succ',
+      Nat.sub_add_cancel (one_le_two.trans (Nat.Prime.two_le hp.out))]
+    rw [mem_powers_iff] at hg
+    obtain ⟨k, hk⟩ := hg
+    rw [← hk, ← pow_mul, mul_comm, pow_mul, pow_eq_iterate f, hf, pow_eq_iterate, iterate_id]
+    rfl
+
+theorem isPGroup_of_powers : @IsPGroup p (powers f) (groupPowers f hf) := by
+  unfold IsPGroup
+  intro ⟨g, hg⟩
+  use 1
+  simp; congr
+  rw [mem_powers_iff] at hg
+  obtain ⟨k, hk⟩ := hg
+  rw [← hk, ← pow_mul, mul_comm, pow_mul, pow_eq_iterate f, hf, pow_eq_iterate, iterate_id]
+  rfl
+
+noncomputable instance : Fintype (MulAction.fixedPoints (powers f) α) :=
+  Fintype.ofFinite (MulAction.fixedPoints (powers f) α)
+
+theorem key (hf : Involutive f) :
+    Fintype.card α ≡ Fintype.card (MulAction.fixedPoints (powers f) α) [MOD 2] := by
+  rw [involutive_iff_iter_2_eq_id] at hf
+  exact @IsPGroup.card_modEq_card_fixedPoints _ _ (_) (isPGroup_of_powers f hf) _ _ _ _ _
+
+end Key
+
+section Involution
+
+open Finset Function Submonoid
+
+variable (k : ℕ) [hk : Fact (4 * k + 1).Prime]
+
+/-- The obvious involution `(x, y, z) ↦ (x, z, y)`. Its fixed points correspond to representations
+of `4 * k + 1` as a sum of two squares. -/
+def obvInvo : Function.End (zagierSet k) := fun ⟨⟨x, y, z⟩, h⟩ => ⟨⟨x, z, y⟩, by
+  unfold zagierSet at *
+  simp_all
+  linarith [h]⟩
+
+lemma obvInvo_apply {p : ℕ × ℕ × ℕ} {hp : p ∈ zagierSet k} :
+    (obvInvo k) ⟨p, hp⟩ = (p.1, p.2.2, p.2.1) := by
+  rfl
+
+theorem involutive_obvInvo : Involutive (obvInvo k) := by
+  unfold Involutive obvInvo zagierSet; simp
+
+theorem sq_add_sq_of_nonempty_fixedPoints
+    (hn : (MulAction.fixedPoints (powers (obvInvo k)) (zagierSet k)).Nonempty) :
+    ∃ a b : ℕ, a ^ 2 + b ^ 2 = 4 * k + 1 := by
+  simp_rw [sq]
+  obtain ⟨⟨⟨x, y, z⟩, he⟩, hf⟩ := hn
+  rw [MulAction.mem_fixedPoints, Subtype.forall] at hf
+  have := hf (obvInvo k) (mem_powers _)
+  apply_fun Subtype.val at this
+  rw [Submonoid.smul_def, End.smul_def] at this
+  unfold obvInvo at this; simp at this
+  unfold zagierSet at he; simp at he
+  use x, (2 * y)
+  rw [this.1, show 4 * y * y = 2 * y * (2 * y) by linarith] at he
+  assumption
+
+/-- The complicated involution, which is shown to have exactly one fixed point `(1, 1, k)`. -/
+def complexInvo : Function.End (zagierSet k) := fun ⟨⟨x, y, z⟩, h⟩ =>
+  ⟨if x + z < y then ⟨x + 2 * z, z, y - x - z⟩ else
+   if 2 * y < x then ⟨x - 2 * y, x + z - y, y⟩ else
+                     ⟨2 * y - x, y, x + z - y⟩, by
+  unfold zagierSet at *
+  simp at h
+  split_ifs with less more <;> simp
+  · rw [Nat.sub_sub]
+    zify [less.le] at h ⊢
+    ring_nf
```

Comment:
```suggestion
```
these shouldn't really be needed, linarith won't care exactly what order the terms are in